Scripts/ScarletMonastery: Rewrite Mograine and Whitemane encounter

Artamedes opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Rewrite the Mograine and Whitemane encounter in Scarlet Monastery
-  also fix the door issue

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #15793 

**Tests performed:** (Does it build, tested in-game, etc.)
Builds and tested ingame

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Artamedes committed:
Scripts/ScarletMonastery: Rewrite Mograine and Whitemane encounter

Artamedes committed:
qq

Artamedes committed:
qq 2

Artamedes committed:
tryfix circleci build

jackpoz commented on src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp:
```diff
@@ -61,321 +53,391 @@ enum Spells
         void Reset() override
         {
-            Initialize();
+            m_CanTalkKill = true;
```
keeping the old Initialize() way reduce code duplication, so these variables are set to their default value only in a function

jackpoz commented:
> * Rewrite the Mograine and Whitemane encounter in Scarlet Monastery

what did the rewrite change ? what did it fix (other than the door issue) ?

jackpoz commented on src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp:
```diff
@@ -61,321 +53,391 @@ enum Spells
+            if (who->GetTypeId() != TYPEID_PLAYER || !m_CanTalkKill)
+                return;
+
+            // Prevent talking more than once - resets every 5 seconds by EVENT_RESET_KILL_TALK
```
what about using a single time_t variable instead ? would that look less complicated than a flag and an event resetting it ?

Artamedes commented:
> > * Rewrite the Mograine and Whitemane encounter in Scarlet Monastery
> 
> what did the rewrite change ? what did it fix (other than the door issue) ? what was the bug with the door ?

The rewrite changed Scarlet Commander Mograine's AI (3976) and High Inquisitor Whitemane's AI (3977). The reason for rewrite is because I took a look at the script, and noticed it was using old codestyle, and is probably an old SD2 script.. So I decided to rewrite it, using new script styles, while documenting a bit, and still keeping mostly old script logic, but changing a few things to make it blizzlike.

For example: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/EasternKingdoms/ScarletMonastery/boss_mograine_and_whitemane.cpp#L324 should've been AOE cast, and she should have also moved near mograine's body and faced him, then started resurrection.

This rewrite probably fixes all issues with this boss, however if some still exist, it can be easily modified and fixed

so the door issue was the door was always open for this encounter, however the door should only open, and stays open after the Mograine first dies (he's later resurrected in the encounter)

Artamedes committed:
review suggestions

Aokromes commented:
tested on http://tcubuntu.northeurope.cloudapp.azure.com/ works as intended.

jackpoz commented:
Thanks for the PR :)
